test/functional/lua/with_spec.lua

```diff
@@ -7,286 +7,1620 @@ local api = n.api
+    local out_buf = exec_lua [[
+      local other_buf, cur_buf = setup_buffers()
+      vim.bo[other_buf].commentstring = '## %s'
+
+      local context = { buf = other_buf, bo = { commentstring = '-- %s' } }
+      local ok, err = pcall(vim._with, context, function() error('Oops buf', 0) end)
+
+      return {
+        ok,
+        err,
+        api.nvim_get_current_buf() == cur_buf,
+        vim.bo[other_buf].commentstring,
+      }
+    ]]
+    eq({ false, 'Oops buf', true, '## %s' }, out_buf)
+
+    local out_win = exec_lua [[
+      local other_win, cur_win = setup_windows()
+      vim.wo[other_win].winblend = 25
+
+      local context = { win = other_win, wo = { winblend = 50 } }
+      local ok, err = pcall(vim._with, context, function() error('Oops win', 0) end)
+
+      return {
+        ok,
+        err,
+        api.nvim_get_current_win() == cur_win,
+        vim.wo[other_win].winblend,
+      }
     ]]
-    )
+    eq({ false, 'Oops win', true, 25 }, out_win)
   end)
 
-  it('layout in current tabpage does not affect windows in others', function()
-    command('tab split')
-    local t2_move_win = api.nvim_get_current_win()
-    command('vsplit')
-    local t2_other_win = api.nvim_get_current_win()
-    command('tabprevious')
-    matches('E36: Not enough room$', pcall_err(command, 'execute "split|"->repeat(&lines)'))
-    command('vsplit')
-
-    exec_lua('vim._with({win = ...}, function() vim.cmd.wincmd "J" end)', t2_move_win)
-    eq({ 'col', { { 'leaf', t2_other_win }, { 'leaf', t2_move_win } } }, fn.winlayout(2))
+  it('handles not supported option', function()
+    local out = exec_lua [[
+      -- Should still restore initial state
+      vim.bo.commentstring = '## %s'
+
+      local context = { o = { commentstring = '-- %s' }, bo = { winblend = 10 } }
+      local ok, err = pcall(vim._with, context, function() end)
+
+      return { ok = ok, err = err, cms = vim.bo.commentstring }
+    ]]
+    eq(false, out.ok)
+    matches('window.*option.*winblend', out.err)
+    eq('## %s', out.cms)
   end)
-end)
 
-describe('vim._with {lockmarks = true}', function()
-  it('is reset', function()
-    local mark = exec_lua [[
-      vim.api.nvim_buf_set_lines(0, 0, 0, false, {"marky", "snarky", "malarkey"})
-      vim.api.nvim_buf_set_mark(0,"m",1,0, {})
-      vim._with({lockmarks = true}, function()
-        vim.api.nvim_buf_set_lines(0, 0, 2, false, {"mass", "mess", "moss"})
-      end)
-      return vim.api.nvim_buf_get_mark(0,"m")
+  it('validates arguments', function()
+    exec_lua [[
+      _G.get_error = function(...)
+        local _, err = pcall(vim._with, ...)
+        return err or ''
+      end
     ]]
-    t.eq(mark, { 1, 0 })
+    local get_error = function(string_args)
+      return exec_lua('return get_error(' .. string_args .. ')')
+    end
+
+    matches('context.*table', get_error("'a', function() end"))
+    matches('f.*function', get_error('{}, 1'))
+
+    local validate_context = function(bad_context, expected_type)
```

Comment:
we are fairly consistent about using `assert_` prefix for tests
```suggestion
    local assert_context = function(bad_context, expected_type)
```